doctor/lib/text_extraction.py

```diff
@@ -0,0 +1,326 @@
+    : is used in NY courts
+    ) is used in many courts
+
+    :param page_text: The text of the first page
+    :return: The page text
+    """
+    for separator in [r"\)", "§", ":"]:
+        matches = list(re.finditer(rf"(.* +{separator} .*\n)", page_text))
+        central_matches = [
+            match
+            for match in matches
+            if 30 <= match.group().rindex(separator[-1]) <= 70
+        ]
+        if len(central_matches) < 3:
+            continue  # Skip this separator if less than 3 matches found
+        # Determine the longest position of the separator
+        longest = max(
+            match.group().rindex(separator[-1]) for match in central_matches
+        )
+        adjust = 0
+        for match in central_matches:
+            match_text = match.group()
+            index = match_text.rindex(separator[-1])
+            location = match.start() + adjust + index
+            # Adjust the page text by adding spaces to align the separators
+            page_text = (
+                page_text[:location]
+                + " " * (longest - index)
+                + page_text[location:]
+            )
+            adjust += longest - index
+        return page_text
+    return page_text
+
+
+def page_needs_ocr(page: pdfplumber.pdf.Page, page_text: str) -> bool:
+    """Does the page need OCR
+
+    :param page:Pdf Plumber Page
+    :param page_text: context extracted from page
+    :return: does page need OCR
+    """
+    if (
+        page_text.strip() == ""
+        or "(cid:" in page_text
+        or page_annotations(page)
+        or page_images(page)
+        or len(page.curves) > 10
+    ):
+        return True
+    return False
+
+
+def convert_pdf_page_to_image(
+    page: pdfplumber.pdf.Page, strip_margin: bool
+) -> Image:
+    """Conver page to image and crop margin if applicable
+
+    :param page:Pdf Plumber
+    :param strip_margin: bool
+    :return: The formatted image
+    """
+    img = page.to_image(resolution=300)
+    _, _, w, h = page.bbox
+    width = w * img.scale
+    height = h * img.scale
+
+    if strip_margin == True:
+        # Because this is OCR - I think its reasonable to crop half the standard
+        # 1 inch margin - this leads to much better results and helps reduce
+        bbox = (
+            (1 / (8.5 * 2) * width),
+            (1 / (11 * 2) * height),
+            (16 / (8.5 * 2) * width),
+            (21 / (11 * 2) * height),
+        )
+        image = img.original.crop(bbox)
+    else:
+        image = img.original
+    return image
+
+
+def ocr_image_to_data(image: Image) -> list[pd.DataFrame]:
+    """Perform OCR on an image to extract data
+
+    Convert the image of the pdf page to OCR data
+    :param image: Pil Image
+    :return: A list of DataFrames, each containing OCR data for a block of text
+    """
+    data_dict = pytesseract.image_to_data(
+        image,
+        config="-c preserve_interword_spaces=1x1 -c tessedit_do_invert=0 --psm 6 -l eng",
```

Comment:
Let's document these either with a link to the docs or by explaining each parameter.

Reply:
done.